### PR TITLE
Add support for more architectures on Linux by doing away with AVX for SIMD.

### PR DESCRIPTION
**Pull request by GustavoSilvera**
## Summary

The strict requirement for x86-64 seemed to be mostly due to the reliance on the `XMM` x86 SIMD intrinsics for SSE (and others like AVX). While they may provide the best performance, it is at a cost of ugly code and architecture limitation (x86-64 only).  

In this work I removed the strict requirement for x86-64 and instead implemented all the SIMD operations using the platform-independent [gcc vector extensions](https://gcc.gnu.org/onlinedocs/gcc/Vector-Extensions.html) (which are [implemented by clang](https://clang.llvm.org/docs/LanguageExtensions.html#vectors-and-extended-vectors)). 

These changes have been tested on Linux `x86-64`  and `aarch64`. Would be great to test on Windows as well!

## Templates

The new implementation allows for generic vector types due to the underlying `__attribute__((ext_vector_type(T)))` so I didn't need to duplicate any code between `F32x4` and `F32x8`, and it is trivial to add `F32x16` or more for instance. 

## Unit tests

Additional unit tests for the SIMD operations is added as well as the matrix inversion function (which makes heavy use of the SIMD operations) which are mostly backwards compatible with the original implementation. 

## Backwards (in)compatibility

Nearly all the existing behaviour has been preserved because the gcc vector extensions already have implemented them in a concise way (ex. `__builtin_shufflevector`, `operator+`, `operator==`) but there were some places that I noticed compatibility differences:

1. [`F32x4::one`](https://github.com/TheNumbat/rpp/blob/b504753fefc489cd927390b7083bd01ac274e498/rpp/impl/simd.cpp#L45-L47) seemed like a bug to me, since it was just calling `_mm_setzero_ps` which is the same behaviour as `F32x4::zero`. This is now implemented with `set1(1)` to set all the initial `f32` values (not bits!) to 1.
2. Getting rid of [`movemask`](https://github.com/TheNumbat/rpp/blob/b504753fefc489cd927390b7083bd01ac274e498/rpp/impl/simd.cpp#L49-L51) and changing the API for [`cmpeq`](https://github.com/TheNumbat/rpp/blob/b504753fefc489cd927390b7083bd01ac274e498/rpp/impl/simd.cpp#L93-L95). Since these were only ever used to [implement vector (in)equality](https://github.com/TheNumbat/rpp/blob/b504753fefc489cd927390b7083bd01ac274e498/rpp/vmath.h#L216-L233), this entire operation could be done more succinctly by using a `reduce()` of the built-in `operator==` comparison. 
3. Changing the endian-ness of the underlying SIMD vector `data`. Since the previous implementation used [`_mm_set_ps`](https://www.intel.com/content/www/us/en/docs/intrinsics-guide/index.html#text=_mm_set_ps) to create an XMM vector, the construction was little-endian `set(e3, e2, e1, e0)` whereas the vector extensions uses a more natural big-endian form `set(e0, e1, e2, e3)` for its constructors. 
4. Using [`sched_yield()`](https://man7.org/linux/man-pages/man2/sched_yield.2.html) instead of [`_mm_pause()`](https://github.com/TheNumbat/rpp/blob/b504753fefc489cd927390b7083bd01ac274e498/rpp/pos/thread_pos.cpp#L35) as a posix-compliant standard that is not x86-64 specific. 

#3 is likely the most wide-reaching potential breakage, but I'd argue it is an improvement if this implementation is to be merged into rpp, otherwise we'd need some code such as [this commit](https://github.com/GustavoSilvera/rpp/commit/76164bea9e29244cba05e8295453feb69bff08c6) to ~~hack-together~~ implement the little-endian mode for the vector extensions. 

## EDIT: update on MSVC:

To preserve compatibility with MSVC on Windows x86-64, I've kept the existing SIMD code for F32x4 and the corresponding vmath matrix inverse. Since it is a simple preprocessor check it should only apply for MSVC while all other platforms will build as long as clang supports the architecture and OS.

### CODE REVIEW

**Comment by TheNumbat** [comment]
Hi, thanks for the PR! Unfortunately, MSVC doesn't implement `ext_vector_type` - this would work under clang on Windows, but I want to maintain MSVC support. 
To support NEON we'll need to #ifdef the SIMD implementations instead. This should be fairly easy since there are only a few operations, and F32x8 can just be deleted for now.

File: test/CMakeLists.txt [hunk 1]
```diff
@@ -11,4 +11,8 @@ foreach(test ${TEST_SOURCES})
     set_target_properties(${testname} PROPERTIES CXX_STANDARD 20 CXX_EXTENSIONS OFF LINKER_LANGUAGE CXX)
     target_link_libraries(${testname} PRIVATE rpp)
     target_include_directories(${testname} PRIVATE ${RPP_INCLUDE_DIRS})
+
+    if (CMAKE_SYSTEM_PROCESSOR MATCHES "(x86)|(X86)|(amd64)|(AMD64)")
```
Comment (TheNumbat) [inline test/CMakeLists.txt:15]: The flag can also be removed here, since optimizing the tests doesn't matter

**Review by TheNumbat — CHANGES_REQUESTED** [review]

**Comment by GustavoSilvera** [comment]
Ah, I didn't realize MSVC doesn't currently implement the [clang or gcc vector extensions](https://developercommunity.visualstudio.com/t/implement-gcc-vector-extensions/351560) :/ bummer. 

What do you think about keeping the x86-64 specific xmm code for win32/64, and leaving the vector extensions for linux builds? So the code is still portable on Linux beyond just x86 and arm64.

File: test/CMakeLists.txt (hunk 1 above)
Comment (GustavoSilvera) [reply test/CMakeLists.txt:15]: True, it just prints an ugly warning when compiling the tests. But can probably be cleaned up with `march=native` or similar.

Comment (TheNumbat) [reply test/CMakeLists.txt:15]: Yeah - it should go away if we delete f32x8

**Comment by TheNumbat** [comment]
I still lean towards no, since the set of `__builtin` operations is quite limited compared to what's available on x86 and arm. Do you have any other architectures in mind? (riscv?)
The swizzling syntax is nice but it can't be used in the MSVC build anyway.

**Comment by GustavoSilvera** [comment]
Well, rpp doesn't seem to really be using any of the fancier features that sse/avx/neon can provide. It's pretty simple usage so far.

I'd argue that the (clang-based) platform independence would be a big benefit so we don't need to hand-code (and hand-test) each SIMD implementation for each architecture that clang supports (even windows 👀).

It would allow for rpp support on Linux x86-64, arm32, arm64, risc-v, PPC, etc. (and MacOS/BSD support too). Not to mention these vector extensions are cross compatible with XMM vectors too.

It should be straightforward to keep the existing xmm implementation compile locked only for MSVC, or even add a build flag to let the users choose.

**Comment by TheNumbat** [comment]
> Well, rpp doesn't seem to really be using any of the fancier features that sse/avx/neon can provide. It's pretty simple usage so far.

Yeah, fair enough---we can use the vector extension in the clang build for now. I may revisit it if I add a richer simd API in the future, though perhaps that should just be a separate library. 

If you rebase this PR, it'll start running CI and you can check that the msvc build works. Since we're removing the arch flags, I also still think we should delete f32x8 (emulating it with sse/neon isn't useful).

**Review by TheNumbat — APPROVED** [review]

**Comment by TheNumbat** [comment]
Ok, looks good - thanks for the contribution. 

I'll see if I can add a macos arm64 CI build for testing arm support.

**Pull request merged**